src/Mvc/Mvc.ViewFeatures/src/Filters/AntiforgeryMiddlewareAuthorizationFilter.cs

```diff
@@ -0,0 +1,37 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using Microsoft.AspNetCore.Antiforgery;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.Core.Filters;
+
+internal partial class AntiforgeryMiddlewareAuthorizationFilter(ILogger<AntiforgeryMiddlewareAuthorizationFilter> logger) : IAsyncAuthorizationFilter
+{
+    internal static string AntiforgeryMiddlewareWithEndpointInvokedKey = "__AntiforgeryMiddlewareWithEndpointInvoked";
+
+    private readonly ILogger _logger = logger;
+
+    public Task OnAuthorizationAsync(AuthorizationFilterContext context)
+    {
+        ArgumentNullException.ThrowIfNull(context);
+
+        if (context.HttpContext.Items.ContainsKey(AntiforgeryMiddlewareWithEndpointInvokedKey))
+        {
+            var antiforgeryValidationFeature = context.HttpContext.Features.Get<IAntiforgeryValidationFeature>();
+            if (antiforgeryValidationFeature is { IsValid: false })
+            {
+                Log.AntiforgeryTokenInvalid(_logger, antiforgeryValidationFeature.Error!.Message, antiforgeryValidationFeature.Error!);
+                context.Result = new AntiforgeryValidationFailedResult();
```

Comment:
This would require new public API since the `AntiforgeryValidationFailedResult` type is defined in the `Mvc.Core` assembly. I'm fine approaching this as a separate optimization for later on.